src/app-layer-smtp.c

```diff
@@ -1396,8 +1478,14 @@ static int SMTPPreProcessCommands(
             line->delim_len = 0;
 
             /* bail if `SMTPProcessRequest` ended the data mode */
-            if ((state->parser_state & SMTP_PARSER_STATE_COMMAND_DATA_MODE) == 0)
+            if ((state->parser_state & SMTP_PARSER_STATE_COMMAND_DATA_MODE) == 0) {
+                if (data_frame == NULL)
+                    data_frame = AppLayerFrameGetLastOpenByType(f, 0, SMTP_FRAME_DATA);
+                if (data_frame) {
+                    data_frame->len = (slice->offset + input->consumed) - data_frame->offset;
```

Comment:
Are we sure we have some contiguous buffer from the original created frame ?

Reply:
yes all data from the open is part of data

Reply:
But what happens if I stream indefinitely some DATA to a smtp server ? Can I get an infinite frame ?
Would Suricata keep a 1Gbyte buffer somewhere ?

Reply:
no, it's a streaming buffer

Reply:
to clarify: it's a pointer into the stream data. So the data kept is managed by the stream engine. Inspection happens in streaming manner, just like with "raw" content stream inspection.

Reply:
Ok, looks magical to me